### PR TITLE
Created ai logic node

### PR DESCRIPTION
**Pull request by MathewMacDougall**
This set of changes outlines the basics for the AI logic node, the node that will be responsible for decision making and path planning.

### CODE REVIEW

**Review by Thunderbots — DISMISSED** [review]

File: src/thunderbots/ai/main.cpp [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+#include "thunderbots_msgs/Ball.h"
+#include "thunderbots_msgs/Field.h"
+#include "thunderbots_msgs/MovePrimitive.h"
+#include "thunderbots_msgs/Team.h"
+
+// Member variables we need to maintain state
+World world;
+// TODO: Use proper abstract classes
+STPHL high_level;
+RRTNav navigator;
```
Comment (henryymliu) [inline src/thunderbots/ai/main.cpp]: nit: can this be Navigator instead of RRTNav?

File: src/thunderbots/ai/main.cpp [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+    // Create subscribers
+    ros::Subscriber field_sub = n.subscribe("backend/field", 1, fieldUpdateCallback);
+    ros::Subscriber ball_sub  = n.subscribe("backend/ball", 1, ballUpdateCallback);
+    ros::Subscriber friendly_team_sub =
+        n.subscribe("backend/friendly_team", 1, friendlyTeamUpdateCallback);
+    ros::Subscriber enemy_team_sub =
+        n.subscribe("backend/enemy_team", 1, enemyTeamUpdateCallback);
+
+    // Initialize variables
```
Comment (henryymliu) [inline src/thunderbots/ai/main.cpp]: nit: variables may not be the best descriptor for these components

File: src/thunderbots/ai/world/field.h [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+     * in metres.
+     *
+     * @param new_field_msg the field message containing the new field information
+     */
+    void updateDimensions(thunderbots_msgs::Field new_field_msg);
+
+    /**
+     * Updates the dimensions of the field. All values should be in metres.
+     *
+     * @param field_length the length of the playing area (along the x-axis)
```
Comment (henryymliu) [inline src/thunderbots/ai/world/field.h]: nit: would be convenient to include units in here as well

File: src/thunderbots/ai/world/robot.h [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+     * @return the predicted position of the robot.
+     */
+    Point position(double time_delta = 0.0) const;
+
+    /**
+     * Get the predicted velocity of the robot at a time relative to the current time.
+     * Using the default
+     * of 0 will give the current velocity of the robot.
+     *
+     * @param time_delta The amount of time in seconds forwards to predict.
```
Comment (henryymliu) [inline src/thunderbots/ai/world/robot.h]: Might want to rephrase this, sounds a little funny.

Comment (henryymliu) [reply src/thunderbots/ai/world/robot.h]: (same with line 73 and 84)

File: src/thunderbots/ai/navigator/navigator.h [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+/**
+ * An abstraction for all navigation operations performed by our AI. The navigator is
+ * responsible for all navigation,
+ * path-planning, and collision-avoidance for our robots.
+ *
+ * This is an Abstract, pure-virtual class. It is meant to define the interace that all
+ * Navigator modules must follow.
+ * Other classes should inherit from this class and implement the methods to create a
+ * useable Navigator class. This
+ * allows us to potentially have multiple different navigators that we can swap out or
```
Comment (henryymliu) [inline src/thunderbots/ai/navigator/navigator.h]: nit: weird line-wrapping

File: src/thunderbots/ai/world/robot.h [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+    /**
+     * Creates a new robot given a pattern id
+     */
+    explicit Robot(const thunderbots_msgs::Robot& robot_msg);
+
+    /**
+     * Updates the state of the robot.
+     *
+     * @param new_position the new position of the robot
```
Comment (henryymliu) [inline src/thunderbots/ai/world/robot.h]: same as field: units would be nice :)

**Review by henryymliu — CHANGES_REQUESTED** [review]
A bunch of nits, but otherwise looks awesome!

File: src/thunderbots/ai/main.cpp (hunk 2 above)
Comment (MathewMacDougall) [reply src/thunderbots/ai/main.cpp]: made `variables` more descriptive

File: src/thunderbots/ai/world/field.h (hunk 3 above)
Comment (MathewMacDougall) [reply src/thunderbots/ai/world/field.h]: units are mentioned in the first line (All units should be in metres). I didn't want to type it out for every line.

File: src/thunderbots/ai/world/robot.h (hunk 6 above)
Comment (MathewMacDougall) [reply src/thunderbots/ai/world/robot.h]: Fixed. Also fixed more weird line wrapping.

**Comment by MathewMacDougall** [comment]
Addressed PR comments and ready for another round of review. I like your convention of using `nit` for stuff :+1:

**Review by henryymliu — APPROVED** [review]

**Pull request merged**